backend/server/db/helpers/users.py

```diff
@@ -0,0 +1,192 @@
+from typing import Optional, Union
+
+import pymongo
+import pymongo.errors
+
+from server.db.mongo.conn import usersNewCOL
+
+from .models import NotSetupUserStorage, PartialUserStorage, UserCoursesStorage, UserDegreeStorage, UserPlannerStorage, UserStorage, YearTerm
+
+# TODO-OLLI(pm): decide if we want to remove type ignores by constructing dictionaries manually
+
+def get_user(uid: str) -> Optional[Union[NotSetupUserStorage, UserStorage]]:
+    res = usersNewCOL.find_one({ 'uid': uid })
+    if res is None:
+        return res
+
+    return UserStorage.model_validate(res) if res["setup"] else NotSetupUserStorage.model_validate(res)
+
+def get_user_degree(uid: str) -> Optional[UserDegreeStorage]:
+    res = get_user(uid)
+    return res.degree if res is not None and res.setup is True else None
+
+def get_user_courses(uid: str) -> Optional[UserCoursesStorage]:
+    res = get_user(uid)
+    return res.courses if res is not None and res.setup is True else None
+
+def get_user_planner(uid: str) -> Optional[UserPlannerStorage]:
+    res = get_user(uid)
+    return res.planner if res is not None and res.setup is True else None
+
+def user_is_setup(uid: str) -> bool:
+    res = usersNewCOL.find_one({ 'uid': uid, 'setup': { "$eq": True } })
+    return res is not None
```

Comment:
I think optimal API is

```
val get_user : uid -> Optional[Union[NotSetupUserStorage, UserStorage]]
val get_user_validate : uid -> Optional[UserStorage]
```

Then, for devs that want access to the `degree`, `planner` fields, etc

```
user = get_user_validated(id)
if user is None: raise ...

# The following fields should now be non-null
use user.program
use user.degree
```

Reply:
good point, thats how it ends up actually working, since everything goes through `get_setup_user` in the routes (which auto throws 403), and the courses/degree/planner is just indexed off of that.

I probably wrote these before i realised that

Reply:
i fixed this